### ISSUE TITLE
Semantics of @example unclear

### ISSUE THREAD

**Issue by rix0rrr** [Microsoft/tsdoc#147]
Hi,

I'm just dipping my toe into TSDoc, so my apologies in advance if this discussion is out of place, or what I'm about to say is common knowledge. Having said that:

There seems to be disagreement about how `@example` blocks are supposed to be interpreted, specifically when using VSCode.

If I add an `@example` block to my doc comment, VSCode responds by nicely syntax-highlighting the following code block:

![image](https://user-images.githubusercontent.com/524162/54130894-8cddf500-4411-11e9-8ac6-ded33375a496.png)

That's amazing, exactly what I want! However, TSDoc tries to parse the code inside the  `@example` block as "regular" docstrings:

![image](https://user-images.githubusercontent.com/524162/54130993-be56c080-4411-11e9-89da-d7147e95f088.png)

Looks like TSDoc wants me to fence the code block with code delimiters, which is not as nice:

![image](https://user-images.githubusercontent.com/524162/54131100-f2ca7c80-4411-11e9-87cc-4bd0f8d5e6fa.png)

(Indentation by 4 spaces, which normally indicates literal code in MarkDown, doesn't work either, TSDoc still tries to parse the code)

**Comment by octogonz** [comment]
From a documentation perspective, I think TSDoc's design is better:

- It allows you to explain your example using English, and possibly present more than one code snippet
- It allows your example to use other code languages (e.g. JavaScript, JSON, XML, etc) and have correct syntax highlighting

> (Indentation by 4 spaces, which normally indicates literal code in MarkDown, doesn't work either, TSDoc still tries to parse the code)

This was an intentional design decision. TSDoc avoids assigning meaning to whitespace, since its easy to mess up inside a `/** */` frame, and unlike Markdown, people writing TSDoc often do not have an immediate rendering feedback loop that helps them predict what it's going to look like on the web site.

I'm not sure what standard VS Code's highlighting is based on. Prior to TSDoc I couldn't find any formal spec for how TypeScript doc comments should be interpreted.

That said, you can see from this thread that `@example` was one of the more dubious tags -- people had some trouble deciding how it should get rendered. Suggestions for improvement would be welcome.

**Comment by rix0rrr** [comment]
Yeah, I see where you're coming from.

It's just this: we're building a class library that we recommend people to write extensions for using VSCode, and we intend to parse the doc comments using TSDoc, and that means we're stuck in an uncomfortable position where the tools we're recommending people to use are going to disagree on what a particular piece of text means. Specifically, the editor is going to apply confusing syntax highlighting to something that's not actually interpreted as code.

(By the way, I just tried mixing descriptive text and fenced code blocks, and VSCode completely loses it when I do that).

I appreciate that while trying to set a new standard there might be a transitionary period during which things are messed up. I just hope that VSCode and TSDoc can align quickly.

Otherwise, to smooth the transition, is there a possibility to have 2 different tags with different interpretations?

**Comment by octogonz** [comment]
> I appreciate that while trying to set a new standard there might be a transitionary period during which things are messed up. I just hope that VSCode and TSDoc can align quickly.

Yeah, we've been planning for a while to set up a conversation with the VS Code owners and improve alignment between the two projects. I've been delaying this because I thought it would be helpful to first write up the spec draft that codifies the TSDoc syntax and philosophy. But the last couple months most of my time in this area has been focuses on shipping API Extractor 7, which will gives us a complete application of TSDoc for a production tool. That should be done in the next week or so, but it's taken much longer than expected. In short, you're right, and we're working towards what you're describing, just rather slower than ideal heheh.

> Otherwise, to smooth the transition, is there a possibility to have 2 different tags with different interpretations?

Sure. The TSDoc parser configuration allows you to define custom tags. So you could make a tag `@richExample` or whatever, which VSCode will not recognize, so it should fallback to recognizing the code fence.

See [this sample code](https://github.com/Microsoft/tsdoc/blob/491e7fdf79271c2e7feba03c98fd148bd014943e/api-demo/src/advancedDemo.ts#L202) for an example.
